flink-dist/src/main/flink-bin/bin/flink-console.sh

```diff
@@ -58,7 +58,34 @@ esac
 
 FLINK_TM_CLASSPATH=`constructFlinkClassPath`
 
-log_setting=("-Dlog4j.configuration=file:${FLINK_CONF_DIR}/log4j-console.properties" "-Dlog4j.configurationFile=file:${FLINK_CONF_DIR}/log4j-console.properties" "-Dlogback.configurationFile=file:${FLINK_CONF_DIR}/logback-console.xml")
+if [ "$FLINK_IDENT_STRING" = "" ]; then
+    FLINK_IDENT_STRING="$USER"
+fi
+
+pid=$FLINK_PID_DIR/flink-$FLINK_IDENT_STRING-$SERVICE.pid
+mkdir -p "$FLINK_PID_DIR"
+# The lock needs to be released after use because this script is started foreground
+command -v flock >/dev/null 2>&1
+flock_exist=$?
+if [[ ${flock_exist} -eq 0 ]]; then
+    exec 200<"$FLINK_PID_DIR"
+    flock 200
+fi
+# Remove the pid file when all the processes are dead
+if [ -f "$pid" ];then
+    all_dead=0
```

Comment:
Why are we re-inventing the wheel and not use the same logic from `flink-daemon.sh`?

`[ $(wc -l < "$pid") -eq 0 ] && rm "$pid"`
or
`[ -z $(tail -n 1 "$pid") ] && rm "$pid"`

Reply:
For `flink-daemon.sh`, it provides the `stop` and `stop-all` command so that we could remove the `pid` file there. However, in `flink-console.sh`, the script could be crashed or killed at any time. We may not have chance to remove it. So i put the clean-up logic when starting a new one. Only when all the processes are dead, we could safely delete the `pid` file.

Reply:
Shouldn't this cleanup be done in a trap of some sort? Like on-exit remove the PID of this process from the file, and then _maybe_ delete it if empty?

Reply:
It is not easy to be done in a `trap` because we need to handle all the signals. The scripts also could be killed -9, then it will not have chance to clean-up. So i suggest to do this before start.